syntax = "proto3";

package todo_app.v1;

import "google/protobuf/empty.proto";
import "resemble/v1alpha1/options.proto";
import "resemble/v1alpha1/tasks.proto";

////////////////////////////////////////////////////////////////////////

service TodoLists {
  option (resemble.v1alpha1.service).state = "TodoListsState";

  rpc AddTodoList(AddTodoListRequest) returns (AddTodoListResponse) {
    option (resemble.v1alpha1.method).transaction = {
    };
  }

  rpc ListTodoLists(ListTodoListsRequest) returns (ListTodoListsResponse) {
    option (resemble.v1alpha1.method).reader = {
    };
  }

  rpc DeleteTodoList(DeleteTodoListRequest) returns (DeleteTodoListResponse) {
    option (resemble.v1alpha1.method).writer = {
    };
  }
}

////////////////////////////////////////////////////////////////////////
// State types.

message TodoListMessage {
  string id = 1;
  string name = 2;
}

message TodoListsState {
  repeated TodoListMessage todolists = 1;
}

////////////////////////////////////////////////////////////////////////
// Request and response types.

message AddTodoListRequest {
  string name = 1;
}

message AddTodoListResponse {
  // The ID of the todolist that was just created.
  string id = 1;
}

message ListTodoListsRequest {}

message ListTodoListsResponse {
  repeated TodoListMessage todolists = 1;
}

message DeleteTodoListRequest {
  string id = 1;
}

message DeleteTodoListResponse {}